pkg/pillar/devicenetwork/dns.go

```diff
@@ -150,21 +154,24 @@ func ResolveWithPortsLambda(domain string,
 						errs = append(errs, err)
 					}
 					if response != nil {
-						resolvedIPsChan <- response
+						select {
+						case <-quit:
+						case resolvedIPsChan <- response:
```

Comment:
Its a great fix.

However, i would add add a timeout to the select statement to prevent the goroutine from blocking indefinitely when trying to send the DNS resolution response to the resolvedIPsChan channel. This will prevent from similar deadlocks in the future.

This timeout could be added on other similar code blocks.
```go
						case <-quit:
						case resolvedIPsChan <- response:
                                                case <-time.After(timeoutDuration):
                                                        return
``` 

Your opinion?

Reply:
We would need this for every `select` and for the `wg.Add(1)`, wouldn't we?

Let's see what will come out of https://github.com/lf-edge/eve/pull/4409#discussion_r1825637446 and perhaps that makes it easier ...

Reply:
I think @cperakis is correct and that this is the root cause of the issue (especially since this PR didn't fix the issue). The channel is unbuffered hence this will block sending on resolvedIPsChan when is more than one response (assuming the channel reader goes away after the first response).

Reply:
But when the reader goes away, the `quit` channel is closed in the defer function and this should unblock this go routine.

Reply:
That doesn't follow normal golang patters - select across a send and a receive on channels is not something I've seen elsewhere.

Reply:
> especially since this PR didn't fix the issue

it's because we tested the version without a fix, by my mistake...

Reply:
Do you mean I should do instead:
```suggestion
						select {
						case resolvedIPsChan <- response:
						default:
						}
```

to make it non-blocking?

Reply:
> That doesn't follow normal golang patters - select across a send and a receive on channels is not something I've seen elsewhere.

It would be nicer to use context.Context with cancel, but actually `ctx.Done()` returns channel so select across send and receive is a common pattern.

Reply:
@milan-zededa how do you feel about [this](https://gist.github.com/uncleDecart/09e3f551f16c7d81fa14acab519adff1) approach?

Reply:
if we want to add timeout in that case we can do it quite easily. Just create a context with timeout :). I'm starting to think it'll make sense to do those changes now

Reply:
> Do you mean I should do instead:
> 
> to make it non-blocking?

That would ensure that it would never block. But it also doesn't wait for the receiver to try to read which AFAIU means that you might get nothing on the channel (that can be avoided by having a length of 1 for the channel).

But if the fix works then I think we should take it and move forward - maybe with some additional comments in the code but no code changes.

Reply:
Yes, that's why the `go test` sometimes failed.
(`go test -count 500 -failfast -race -v . -run=TestResolveWithPortsLambda$` to reproduce)

Making the channel of size `1` fixed it.

Reply:
@uncleDecart I think you should not need line 91: `cancel() // Cancel remeaning goroutines`
The context is canceled by defer of `resolveDNS`. 
Regarding timeout: we already have 30 seconds timeout inside `resolve()`. So I'm not sure if we want/need to have a global timeout for the entire `ResolveWithPortsLambda` function.